### PR TITLE
Refactor Tabs

### PR DESCRIPTION
**Pull request by dannyroberts**
🐡 
Probably the most important changes are:
- got rid of the concept of subtabs (UI had been inadvertently removed 6+ months ago)
- vastly simplified active tab election—UITabs now declare what url prefixes they're claiming, and there's no attempt at inference beyond that
  - in order to make it easy for myself to make this change without screwing it up, in DEBUG mode if a class doesn't have `url_prefix_formats` set, it goes through your urls (like it was doing live before) and infers the prefixes, and suggests them to you in an exception message.
- as a result of better tab election, a simple page went from ~9 sec to 1-2 (this was an apps setting page), and in general the header by itself now takes well under a second to load, where it had taken a lot longer before

### CODE REVIEW

**Comment by dannyroberts** [comment]
@snopoke buddy, cc: @biyeun @esoergel who I mentioned this to

**Comment by dannyroberts** [comment]
I've tested a lot of different situations already, but I'm putting it up on staging and sleeping on it. I'll do a final pass tomorrow and remove the QA label.

**Comment by czue** [comment]
a little :night_with_stars: :gift: from the :racehorse: :crown: !

**Comment by snopoke** [comment]
First pass looks good. Build error is real:
```
======================================================================
ERROR: test_add_repeater (corehq.apps.domain.tests.test_views.TestDomainViews)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/apps/domain/tests/test_views.py", line 68, in test_add_repeater
    response = self.client.post(post_url, {'url': forwarding_url}, follow=True)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 484, in post
    response = self._handle_redirects(response, **extra)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 622, in _handle_redirects
    response = self.get(url.path, QueryDict(url.query), follow=False, **extra)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 470, in get
    **extra)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 286, in get
    return self.generic('GET', path, secure=secure, **r)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 358, in generic
    return self.request(**r)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 440, in request
    six.reraise(*exc_info)
  File "/vendor/lib/python2.7/site-packages/django/core/handlers/base.py", line 111, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/vendor/lib/python2.7/site-packages/django/views/generic/base.py", line 69, in view
    return self.dispatch(request, *args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/utils/decorators.py", line 29, in _wrapper
    return bound_func(*args, **kwargs)
  File "/mnt/corehq/apps/domain/decorators.py", line 93, in _inner
    return view_func(req, domain_name, *args, **kwargs)
  File "/mnt/corehq/apps/domain/decorators.py", line 302, in _inner
    return view_func(request, domain_name, *args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/utils/decorators.py", line 25, in bound_func
    return func.__get__(self, type(self))(*args2, **kwargs2)
  File "/mnt/corehq/apps/style/decorators.py", line 23, in _wrapped
    return view_func(class_based_view, request, *args, **kwargs)
  File "/mnt/corehq/apps/domain/views.py", line 2217, in dispatch
    return super(BaseProjectSettingsView, self).dispatch(request, *args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/utils/decorators.py", line 29, in _wrapper
    return bound_func(*args, **kwargs)
  File "/mnt/corehq/apps/domain/decorators.py", line 93, in _inner
    return view_func(req, domain_name, *args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/utils/decorators.py", line 25, in bound_func
    return func.__get__(self, type(self))(*args2, **kwargs2)
  File "/mnt/corehq/apps/domain/views.py", line 200, in dispatch
    return super(LoginAndDomainMixin, self).dispatch(*args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/views/generic/base.py", line 87, in dispatch
    return handler(request, *args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/views/generic/base.py", line 154, in get
    context = self.get_context_data(**kwargs)
  File "/mnt/corehq/apps/hqwebapp/views.py", line 749, in get_context_data
    context.update(self.main_context)
  File "/mnt/corehq/apps/domain/views.py", line 287, in main_context
    project=self.request.project
TypeError: __init__() got multiple values for keyword argument 'domain'
```

**Comment by dannyroberts** [comment]
✅

**Comment by snopoke** [comment]
:+1: after QA

**Comment by dannyroberts** [comment]
Clicked around on staging, everything looks great! (And dare I say... a little snappier just maybe?)

**Comment by snopoke** [comment]
Great work Danny!

File: corehq/apps/reports/dispatcher.py [hunk 1]
```diff
@@ -174,16 +174,12 @@ def allowed_renderings(cls):
...
-    def navigation_sections(cls, context):
-        request = context.get('request')
-        domain = context.get('domain') or getattr(request, 'domain', None)
+    def navigation_sections(cls, request, domain):
         project = getattr(request, 'project', None)
         couch_user = getattr(request, 'couch_user', None)
-        
         nav_context = []
 
         dispatcher = cls()  # uhoh
```
Comment (benrudolph) [inline corehq/apps/reports/dispatcher.py]: i hate this comment haha. makes me feel like something bad is going to happen

File: corehq/tabs/templatetags/menu_tags.py [hunk 2]
```diff
@@ -60,18 +67,13 @@ def format_sidebar(context):
...
         # set is_active on active sidebar item by modifying nav by reference
         # and see if the nav needs a subnav for the current contextual item
         for section_title, navs in sections:
             for nav in navs:
-                full_path = request.get_full_path()  # The path of the URL after the domain
-                absolute_uri = request.build_absolute_uri()  # The full uri {scheme}{host}{path}
-                if (full_path.startswith(nav['url']) or
-                   _strip_scheme(absolute_uri).startswith(_strip_scheme(nav['url']))):
+                full_path = request.get_full_path()
+                if path_starts_with_url(full_path, nav['url']):
```
Comment (benrudolph) [inline corehq/tabs/templatetags/menu_tags.py]: yay

File: corehq/tabs/uitab.py [hunk 3]
```diff
@@ -70,25 +70,24 @@ def sidebar_items(self):
...
-        real_is_viewable and wrapped in a try block that returns False in the
+        should_show and wrapped in a try block that returns False in the
         case of an AttributeError for any of those variables.
 
         """
         raise NotImplementedError()
 
-    @property
     @memoized
-    def real_is_viewable(self):
```
Comment (benrudolph) [inline corehq/tabs/uitab.py]: hahh

**Comment by benrudolph** [comment]
looks good and staging _does_ feel snappier! 🍻 here's to a good new relic report next week

**Comment by czue** [comment]
:clap:

**Comment by biyeun** [comment]
:boom: 🎉

**Pull request merged**